drafts/draft-0005.md

```diff
@@ -0,0 +1,142 @@
+## Preamble
+
+```
+SEP: <to be assigned>
+Title: Data Entry Namespaces
+Author: Mister.Ticot <mister.ticot@cosmic.plus>
+Status: Draft
+Created: 2018-10-26
+```
+
+## Simple Summary
+
+Defines a standard syntax to represent namespaces in data entries.
+
+## Abstract
+
+This SEP provides a standard way to represent
+[namespaces](https://en.wikipedia.org/wiki/Namespace) into Stellar accounts
+data entries, and provides guideline about how to parse them and to handle
+edge cases.
+
+## Motivation
+
+As a generic mechanism to store key/value pairs on the ledger, data entries 
+accept a very wide range of use cases. If each actor use it following its own 
+syntax, it will lead to a confusing experience and ultimatelly to naming 
+conflicts.
+
+Introducing namespaces allow to hierarchize and organize data, in such a way 
+that sorting the key/value entries will result in a readable output even when 
+using them for various purpose on the same account. It also helps to prevent
+name clash.
+
+The present SEP ensure that all other SEP that will use data entries will do so 
+in a consistent way. It provides guidelines ontop of which generic tool to 
+handle data entries can be created.
+
+## Specification
+
+Namespace hierarchy is encoded in the key using one or more terms separated by 
+dots. Terms are made of lowcase letters, undescore and numbers and musn't start
+with a number.
+
+**Examples**
+
+```
+conf.two_factor
+conf.multisig.server
+conf.multisig.network
+conf.multisig.key
+profile.name
+profile.language
+wallet.btc
+wallet.eth
+```
+
+* **Term REGEXP:** `[_a-z][_a-z0-9]*`
+* **Key REGEXP:** `[_a-z][_a-z0-9]*(\.[_a-z][_a-z0-9]*)*`
+
+When parsing the data tree in a given programming language, keys that doesn't 
+respect this syntax must be ignored and valid entries must be hierarchized in 
+an adequate data structure. If a term is used both as an identifier and a 
+namespace, a warning must be thrown by the parser and the identifier value must 
+be ignored.
+
+
+**Example: same data tree in JSON**
+
+```js
+{
+  conf: {
+    two_factor: ...,
+    multisig: {
+      server: ...,
+      network: ...,
+      key: ...
+    }
+  },
+  profile: {
+    name: ...,
+    language: ...
+  },
+  wallet: {
+    btc: ...,
+    eth: ...
+  }
+}
+```
+
+
+## Rationale
+
+**Syntax**
+
+The namespace syntax had to be familiar, and terms have to be written in a way
+that is valid accross a broad range of programming languages.
+
+This protocol uses JavaScript syntax because it is the language the most 
+developers uses. It is also familiar to anybody using JSON, which is a format 
+widely used to pass data, and the one already used by horizon API.
+
+Parsing the data tree must lead to the same namespace/key/value structure 
+regardless of the programming language being used. This is the reason why term 
+syntax must be restricted, as any unusual character could lead to errors in 
```

Comment:
If we're thinking about this in terms of JSON, then I think that you should allow anything that can be a valid JSON key -- which is everything.

Reply:
But then why would the name matter at all for the programming language? We're just talking about strings. When you mention "key" it implies a JSON key.

Reply:
I guess i'm stepping around my main opinion here: that I don't think namespace keys should have _any_ restrictions in them, and that this sep could then boil down to a suggestion of: "use dots (.) to group hierarchically related data under a namespace".

Reply:
> But then why would the name matter at all for the programming language?

I don't know how to put more clearly than I already did. You won't be able to parse the structure if key names are not valid in a given language.

>  When you mention "key" it implies a JSON key

This is only your interpretation. I'm using key in reference of a key/value pair which's a generic concept.

If I understand your position, you want to remove the feature of being able to parse the data tree into a structure, at the cost of having each one having to code its own solution to handle the tree; which strange because you used the argument of "cost of implementation" a couple of days ago. Do you realize there are programming languages in which it won't be a trivial task?

Another drawback I can see in this proposal is that we loose in cleanness, as we could end up with a tree like:

```
profile.name
Profile.myPubKey
PROFILE.iMaGr34Th4cKeR
```

Enforcing lowcase actually ensure an item goes under the intended namespace.

So, I'd say the cost of removing character restrictions is pretty high. It is OK to have opinions about what the SEP should be, but if there's no rational motivation and it only have drawbacks why should we follow it?

Reply:
> You won't be able to parse the structure if key names are not valid in a given language.

Parse into what? I'm totally lost here about here now since we're not talking JSON and we're not talking any specific languages so I have no context besides JSON. Any language that can use strings can build a tree out them.

As far as lowercase vs uppercase -- sure, it's cool to define whether "CONFIG" and "config" are the same namespace, but since we can't actually restrict what gets in the data entries I don't think it makes sense to restrict it here as well beyond saying "namespaces are considered case insensitive".

Reply:
By the way saying that I could make design choices for their "coolness" is kind of insulting.

Reply:
> it is easier to reference conf.foo than conf.['foo*.!']

But you wouldn't need to do that unless you you had a namespace called 'foo*!' that you cared about accessing. 

The benefit from removing the restrictions is that it simplifies the SEP and gets at the core bits that we should really care about: that we are encouraging people to group things using dots and setting the stage for "well known" namespaces. As it stands, my opinion is that the restrictions are adding some complexity for no benefit.

Reply:
Ok, so the benefits are:

* **It is by far the simpler way to make it case insensitive**
* It is also the simpler way to incentive a "clean" usage of namespaces
* In some languages, the easiest / most efficient way to build the data tree may actually comes with restrictions (for example in LISP languages you would use symbols, which comes with restrictions). As we don't know about all languages that may end up interfacing with Stellar, it is easier to solve it with tight rules.
* It is easy to handle with a simple regexp

The disadvantages is:

* It takes two lines to define the REGEXP and a few other to explain the rational behind this choice.

Now the benefits of removing restrictions:

* You can use keys like `=(G_G)=`
* You can use keys like `I love spaces!`

The disadvantages:

* You break the sorted raw display because of higher case letters
* It requires more code & computation to ensure case insensitiveness; So it makes it harder to implement.
* It is actually not clear which character you can use or not, as Stellar keys comes with its own restrictions for keys, but anyway there will still be a restriction of same kind.
* The advantages are not real advantages